### PR TITLE
feat: add 0.x upgrade guide for deprecations

### PR DESCRIPTION
**Pull request by majorgreys**
Add `DD_TRACE_RAISE_DEPRECATIONWARNING` and `ddtrace.internal.utils.deprecations.DDTraceDeprecationWarning`. Release note for next minor version on 0.x will also include upgrade steps in preparation for v1.0.0. The docs for deprecation warnings are also updated.

### CODE REVIEW

File: ddtrace/__init__.py [hunk 1]
```diff
@@ -13,6 +13,11 @@
 # a global tracer instance with integration settings
 tracer = Tracer()
 
+
+class DDTraceDeprecationWarning(DeprecationWarning):
```
Comment (brettlangdon) [inline ddtrace/__init__.py:17]: you are probably going to get hit with circular imports here, might need to move to `ddtrace.internal.*` somewhere and expose publicly here

File: ddtrace/contrib/flask/middleware.py [hunk 2]
```diff
@@ -27,8 +28,8 @@ def _normalize_resource(resource):
     return compat.to_unicode(resource).lower()
 
 
-class TraceMiddleware(object):
-    @remove(message="Use patching instead (see the docs).", removal_version="1.0.0")
+class TraceMiddleware(object):DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/flask/middleware.py:31]:
```suggestion
class TraceMiddleware(object):
```

File: ddtrace/contrib/celery/task.py [hunk 3]
```diff
@@ -23,9 +26,10 @@ def unpatch_task(task):
     via unpatch() API. This API is now a no-op implementation so it doesn't
     affect instrumented tasks.
     """
-    deprecate(
+    deprecate(DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/celery/task.py:29]:
```suggestion
    deprecate(
```

File: ddtrace/contrib/celery/task.py [hunk 4]
```diff
@@ -8,8 +10,9 @@ def patch_task(task, pin=None):
     enables instrumentation on all tasks.
     """
     deprecate(
-        "ddtrace.contrib.celery.patch_task is deprecated",
+        "ddtrace.DDTraceDeprecationWarning_task is deprecated",
```
Comment (brettlangdon) [inline ddtrace/contrib/celery/task.py:13]:
```suggestion
        "ddtrace.contrib.celery.patch_task is deprecated",
```

File: ddtrace/contrib/cassandra/session.py [hunk 5]
```diff
@@ -277,8 +278,8 @@ def _sanitize_query(span, query):
 
 #
 # DEPRECATED
-#
-@remove(message="Use patching instead (see the docs).", removal_version="1.0.0")
+#DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/cassandra/session.py:281]:
```suggestion
#
```

File: ddtrace/contrib/mongoengine/patch.py [hunk 6]
```diff
@@ -14,8 +16,8 @@ def patch():
 
 def unpatch():
     setattr(mongoengine, "connect", _connect)
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/mongoengine/patch.py:19]:
```suggestion

```

File: ddtrace/contrib/mysql/tracers.py [hunk 7]
```diff
@@ -1,8 +1,10 @@
 import mysql.connector
 
-from ...vendor.debtcollector.removals import remove
+from ddtrace import V1DeprecationWarning
 
+from ...vendor.debtcollector.removals import remove
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/mysql/tracers.py:6]:
```suggestion
```

File: ddtrace/contrib/psycopg/connection.py [hunk 8]
```diff
@@ -6,15 +6,17 @@
...
 
+from ddtrace import V1DeprecationWarning
+
 from ...constants import SPAN_MEASURED_KEY
 from ...ext import SpanTypes
 from ...ext import db
 from ...ext import net
 from ...ext import sql
 from ...vendor.debtcollector.removals import remove
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/psycopg/connection.py:17]:
```suggestion

```

File: ddtrace/contrib/pymongo/patch.py [hunk 9]
```diff
@@ -39,8 +40,8 @@ def unpatch():
     unpatch_pymongo_module()
     setattr(pymongo, "MongoClient", _MongoClient)
 
-
-@remove(message="Use patching instead (see the docs).", removal_version="1.0.0")
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/pymongo/patch.py:43]:
```suggestion

```

File: ddtrace/contrib/pymysql/tracers.py [hunk 10]
```diff
@@ -1,8 +1,10 @@
 import pymysql.connections
 
-from ...vendor.debtcollector.removals import remove
+from ddtrace import V1DeprecationWarning
 
+from ...vendor.debtcollector.removals import remove
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/pymysql/tracers.py:6]:
```suggestion

```

File: ddtrace/contrib/redis/tracers.py [hunk 11]
```diff
@@ -1,17 +1,19 @@
 from redis import StrictRedis
 
+from ddtrace import V1DeprecationWarning
+
 from ...vendor.debtcollector.removals import remove
 
 
 DEFAULT_SERVICE = "redis"
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/redis/tracers.py:9]:
```suggestion
```

File: ddtrace/contrib/redis/tracers.py [hunk 12]
```diff
@@ -1,17 +1,19 @@
...
 
-
-@remove(message="Use patching instead (see the docs).", removal_version="1.0.0")
+@remove(message="Use patching instead (see the docs).", category=V1DeprecationWarning, removal_version="1.0.0")
 def get_traced_redis(ddtracer, service=DEFAULT_SERVICE, meta=None):
     return _get_traced_redis(ddtracer, StrictRedis, service, meta)
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/redis/tracers.py:14]:
```suggestion

```

File: ddtrace/contrib/requests/legacy.py [hunk 13]
```diff
@@ -24,8 +26,9 @@ def _distributed_tracing_setter(self, value):
     of the Tracer.
     """
     deprecate(
-        "client.distributed_tracing is deprecated",
+        "client.dDDTraceDeprecationWarnings deprecated",
```
Comment (brettlangdon) [inline ddtrace/contrib/requests/legacy.py:29]:
```suggestion
        "client.distributed_tracing deprecated",
```

File: ddtrace/contrib/sqlite3/connection.py [hunk 14]
```diff
@@ -1,8 +1,10 @@
 from sqlite3 import Connection
 
-from ...vendor.debtcollector.removals import remove
+from ddtrace import V1DeprecationWarning
 
+from ...vendor.debtcollector.removals import remove
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/sqlite3/connection.py:6]:
```suggestion
```

File: ddtrace/contrib/util.py [hunk 15]
```diff
@@ -1,12 +1,15 @@
...
+from ddtrace import V1DeprecationWarning
+
 from ..internal.utils.importlib import func_name
 from ..internal.utils.importlib import module_name
 from ..internal.utils.importlib import require_modules
 from ..vendor.debtcollector.removals import removed_module
 
 
-removed_module(
+removed_modulDDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/contrib/util.py:10]:
```suggestion
removed_module(
```

File: ddtrace/ext/errors.py [hunk 16]
```diff
@@ -17,17 +18,18 @@
...
+    message="DDTraceDeprecationWarningE, and ERROR_STACK.",
+    category=V1DeprecationWarning,
     removal_version="1.0.0",
 )
 
 # shorthand for ERROR constants to be removed in v1.0-----^
 MSG = ERROR_MSG
 TYPE = ERROR_TYPE
 STACK = ERROR_STACK
+DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/ext/errors.py:30]:
```suggestion

```

File: ddtrace/ext/errors.py [hunk 17]
```diff
@@ -17,17 +18,18 @@
 removed_module(
     module="ddtrace.ext.errors",
     replacement="ddtrace.constants",
-    message="ERROR_MSG, ERROR_TYPE, and ERROR_STACK.",
+    message="DDTraceDeprecationWarningE, and ERROR_STACK.",
```
Comment (brettlangdon) [inline ddtrace/ext/errors.py:21]:
```suggestion
    message="ERROR_MSG, ERROR_TYPE, and ERROR_STACK.",
```

File: ddtrace/__init__.py (hunk 1 above)
Comment (majorgreys) [reply ddtrace/__init__.py:17]: Hm, but if we do that wouldn't the class path to the category change?

Comment (mabdinur) [reply ddtrace/__init__.py:17]: We can move it to `ddtrace.internal.utils.deprecation.py`. We had something similar there before

**Review by mabdinur — DISMISSED** [review]

File: ddtrace/__init__.py [hunk 18]
```diff
@@ -1,11 +1,16 @@
+import os
+import warnings as py_warnings
```
Comment (brettlangdon) [inline ddtrace/__init__.py:2]: A comment here would be great

File: ddtrace/__init__.py [hunk 19]
```diff
@@ -1,11 +1,16 @@
...
 from ._monkey import patch  # noqa: E402
 from ._monkey import patch_all
+from .internal.utils.formats import asbool
 from .pin import Pin  # noqa: E402
 from .settings import _config as config  # noqa: E402
 from .span import Span  # noqa: E402
 from .tracer import Tracer  # noqa: E402
 from .vendor.debtcollector.removals import remove
 from .version import get_version
+from .warnings import DDTraceDeprecationWarning
```
Comment (brettlangdon) [inline ddtrace/__init__.py:13]: we can also consider renaming our warnings module to something else as well

File: ddtrace/warnings.py [hunk 20]
```diff
@@ -0,0 +1,2 @@
+class DDTraceDeprecationWarning(DeprecationWarning):
```
Comment (brettlangdon) [inline ddtrace/warnings.py:1]: docstring?

File: releasenotes/notes/add-deprecation-warning-category-1750b81349935713.yaml [hunk 21]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add ``DD_TRACE_RAISE_DEPRECATIONWARNING`` to raise an exception for warnings with the category :py:class:`ddtrace.warning.DDTraceDeprecationWarning`.
```
Comment (brettlangdon) [inline releasenotes/notes/add-deprecation-warning-category-1750b81349935713.yaml:4]:
```suggestion
    Add ``DD_TRACE_RAISE_DEPRECATIONWARNING`` to raise an exception for warnings with the category :py:class:`ddtrace.warnings.DDTraceDeprecationWarning`.
```

Comment (brettlangdon) [reply releasenotes/notes/add-deprecation-warning-category-1750b81349935713.yaml:4]:
```suggestion
    Add ``DD_TRACE_RAISE_DEPRECATIONWARNINGS`` to raise an exception for warnings with the category :py:class:`ddtrace.warning.DDTraceDeprecationWarning`.
```

?

File: releasenotes/notes/add-ddtrace-deprecation-category-f435cd2447d97899.yaml [hunk 22]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add ``ddtrace.DDTraceDeprecationWarning``.
```
Comment (brettlangdon) [inline releasenotes/notes/add-ddtrace-deprecation-category-f435cd2447d97899.yaml:4]:
```suggestion
    Add ``ddtrace.warnings.DDTraceDeprecationWarning``.
```

>

File: ddtrace/compat.py [hunk 23]
```diff
@@ -1,3 +1,5 @@
+from ddtrace.warnings import DDTraceDeprecationWarning
```
Comment (mabdinur) [inline ddtrace/compat.py:1]:
```suggestion
from ddtrace.internal.utils.deprecation import DDTraceDeprecationWarning
```

File: releasenotes/notes/add-ddtrace-deprecation-category-f435cd2447d97899.yaml [hunk 24]
```diff
@@ -0,0 +1,16 @@
+---
+prelude: >
+  In preparation for the upcoming v1.0.0 release, we recommend upgrade to ``ddtrace>=0.59.0,<1.0.0`` and configuring `warning filters <https://docs.python.org/3.8/library/warnings.html#warning-filter>`_.
+
+  Users of pytests can add a filter for ``ddtrace.DDTraceDeprecationWarning``::
+    pytest -W "error::ddtrace.DDTraceDeprecationWarning" tests.py
+
+
+  Otherwise, set the environment variable ``DD_TRACE_RAISE_DEPRECATIONWARNING``::
```
Comment (brettlangdon) [inline releasenotes/notes/add-ddtrace-deprecation-category-f435cd2447d97899.yaml:9]: Should we explicitly state "to turn deprecation warnings into exceptions"?

Wonder if this is clear enough that your app will crash with this setting enabled?

**Review by mabdinur — DISMISSED** [review]

File: ddtrace/internal/utils/deprecations.py [hunk 25]
```diff
@@ -0,0 +1,8 @@
+class DDTraceDeprecationWarning(DeprecationWarning):
+    pass
+
+
+# Override module to simplify adding warning filters by querying for
+# ddtrace.DDTraceDeprecationWarning but not have to expose this in the
+# public API.
+DDTraceDeprecationWarning.__module__ = "ddtrace"
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/utils/deprecations.py:8]: I think we could just declare this in `ddtrace.__init__` since we import it there and expose it anyway

**Review by Kyle-Verhoog — DISMISSED** [review]
Just a thought but lgtm as-is

File: ddtrace/internal/utils/deprecations.py (hunk 25 above)
Comment (mabdinur) [reply ddtrace/internal/utils/deprecations.py:8]: That approach results in circular imports. We use ``DDTraceDeprecationWarning` in ``ddtrace.tracer`` and ``ddtrace.config``

Comment (Kyle-Verhoog) [reply ddtrace/internal/utils/deprecations.py:8]: Ah probably worth documenting that as well here then

**Review by mabdinur — APPROVED** [review]

File: ddtrace/__init__.py [hunk 26]
```diff
@@ -24,11 +29,15 @@
 ]
 
 
-@remove(removal_version="1.0.0")
+if asbool(os.getenv("DD_TRACE_RAISE_DEPRECATIONWARNING")):
+    warnings.filterwarnings(action="error", category=DDTraceDeprecationWarning)
```
Comment (brettlangdon) [inline ddtrace/__init__.py:33]:
```suggestion
if asbool(os.getenv("DD_TRACE_RAISE_DEPRECATIONWARNING")):
    warnings.filterwarnings(action="error", category=DDTraceDeprecationWarning)
```

I can add as a follow-up PR, `DD_TRACE_LOG_DEPRECATIONWARNING` could be a nice to have too, `warnings.filterwarnings(action="default", category=DDTraceDeprecationWarning)`

**Review by brettlangdon — APPROVED** [review]

**Comment by brettlangdon** [comment]
@Mergifyio backport 0.59

**Pull request merged**